src/RuntimeGeneratedFunctions.jl

```diff
@@ -20,56 +36,59 @@ differences are:
 You need to use `RuntimeGeneratedFunctions.init(your_module)` a single time at
 the top level of `your_module` before any other uses of the macro.
 
+If provided, `context_module` is module in which symbols within
+`function_expression` will be looked up. By default this is module in which
+`@RuntimeGeneratedFunction` is expanded.
+
 # Examples
 ```
 RuntimeGeneratedFunctions.init(@__MODULE__) # Required at module top-level
 
 function foo()
     expression = :((x,y)->x+y+1) # May be generated dynamically
-    f = RuntimeGeneratedFunction(@__MODULE__, expression)
+    f = @RuntimeGeneratedFunction(expression)
     f(1,2) # May be called immediately
 end
 ```
 """
-struct RuntimeGeneratedFunction{argnames,moduletag,id} <: Function
-    body::Expr
-    function RuntimeGeneratedFunction(mod::Module, ex)
-        if !isdefined(mod, _tagname)
-            error("""You must use `RuntimeGeneratedFunctions.init(@__MODULE__)` at module
-                     top level before using runtime generated functions""")
-        end
-        moduletag = getfield(mod, _tagname)
-        def = splitdef(ex)
-        args, body = normalize_args(def[:args]), def[:body]
-        id = expr_to_id(body)
-        cached_body = _cache_body(moduletag, id, body)
-        new{Tuple(args),moduletag,id}(cached_body)
-    end
+macro RuntimeGeneratedFunction(code)
+    _RGF_constructor_code(:(@__MODULE__), esc(code))
+end
+macro RuntimeGeneratedFunction(context_module, code)
+    _RGF_constructor_code(esc(context_module), esc(code))
 end
 
-
-macro RuntimeGeneratedFunction(ex)
-    Base.depwarn("`@RuntimeGeneratedFunction(ex)` is deprecated, use `RuntimeGeneratedFunction(@__MODULE__, ex)` instead.", :RuntimeGeneratedFunction)
+function _RGF_constructor_code(context_module, code)
     quote
-        RuntimeGeneratedFunction(@__MODULE__, $(esc(ex)))
+        code = $code
+        cache_module = @__MODULE__
```

Comment:
Hi @c42f , could I ask if there's a specific reason for setting the `cache_module` to be `@__MODULE__` here, rather than allowing the user to provide it? This unfortunately regresses the `RGFPrecompConsumerTest` I had in #19 (though it wasn't merged in). I've had to work around this in SciML/ModelingToolkit.jl#769 by manually calling the `RuntimeGeneratedFunction` constructor instead.

Reply:
No offense taken!

The thought was that we were protecting users from accidentally breaking the precompilation cache mechanism (see https://github.com/SciML/RuntimeGeneratedFunctions.jl/issues/21) — originally the function version of `RuntimeGeneratedFunction` wasn't meant to be used as a constructor; the thought was that the macro would be the official constructor.

However, this doesn't cover all use cases as you noticed. So I've created #26 to allow you to set what you need without messing with the tag struct (which is definitely meant to be internal).

Reply:
Thanks a lot mate for your explanation and the work in #26, that'll help simplify the workaround in ModelingToolkit for sure!